x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/TemplateParamValidator.java

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.application.search;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import com.networknt.schema.JsonSchema;
+import com.networknt.schema.JsonSchemaFactory;
+import com.networknt.schema.SpecVersion;
+import com.networknt.schema.ValidationMessage;
+
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.ValidationException;
+import org.elasticsearch.common.bytes.BytesArray;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.xcontent.ToXContentObject;
+import org.elasticsearch.xcontent.XContentBuilder;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+
+/**
+ * Contains the template parameter validation, as a JSON Schema. It validates that the JSON schema
+ * is valid, and will apply it to the query parameters
+ */
+public class TemplateParamValidator implements ToXContentObject, Writeable {
+
+    private static final SpecVersion.VersionFlag SCHEMA_VERSION = SpecVersion.VersionFlag.V7;
+    private static final JsonSchemaFactory SCHEMA_FACTORY = JsonSchemaFactory.getInstance(SCHEMA_VERSION);
+    private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();
+    private static final JsonSchema META_SCHEMA = SCHEMA_FACTORY.getSchema(
+        TemplateParamValidator.class.getResourceAsStream("json-schema-draft-07.json")
+    );
+    private static final String PROPERTIES_NODE = "properties";
+
+    private final JsonSchema jsonSchema;
+
+    public TemplateParamValidator(StreamInput in) throws IOException {
+        this(in.readString());
+    }
+
+    public TemplateParamValidator(String dictionaryContent) throws ValidationException {
+        try {
+            // Create a new Schema with "properties" node based on the dictionary content
+            final ObjectNode schemaJsonNode = OBJECT_MAPPER.createObjectNode();
+            schemaJsonNode.set(PROPERTIES_NODE, OBJECT_MAPPER.readTree(dictionaryContent));
+            validateWithSchema(META_SCHEMA, schemaJsonNode);
+
+            this.jsonSchema = SCHEMA_FACTORY.getSchema(schemaJsonNode);
```

Comment:
when the dictionary is:
```
        "dictionary": {
            "field_name": {
                "type": "string"
            },
            "field_value": {
                "type": "string"
            }
        }
```

this will create a JSON schema like:
```
{
  "type": "object",
  "properties": {
            "field_name": {
                "type": "string"
            },
            "field_value": {
                "type": "string"
            }
  }
}
```

with this pattern, it will not be possible to specify which parameters are required.
maybe we should re-evaluate how we will use specify the JSON schema

Reply:
> I noticed that we can create a dictionary like:
>     "dictionary": {
>       "lalala": { "lalala": "lalala" }
>     }
> ```
> does this actually generate a valid JSON schema?

I'm afraid so, as there is no `additionalProperties: false` in [the JSON schema validation](https://json-schema.org/draft-07/schema) used to validate the JSON Schema 😢

Reply:
> with this pattern, it will not be possible to specify which parameters are required.
maybe we should re-evaluate how we will use specify the JSON schema

That is correct. It is not possible as well to define custom types.

We could add the full JSON schema and be transparent about what we're using. It would look like:

```json
{
    "indices": ["test1"],
    "template": {
        "script": {
            "source": {
                "query": {
                    "term": {
                        "{{field_name}}": "{{field_value}}"
                    }
                }
            },
            "params": {
                "field_name": "hello",
                "field_value": "world"
            }
        },
        "params_schema": {
            "definitions": {
                "person": {
                    "type": "object",
                    "required": ["first_name", "last_name", "age"],
                    "properties": {
                        "first_name": {"type": "string"},
                        "last_name": {"type": "string"},
                        "age": {"type": "integer"}
                    }
                }
            },
            "properties": {
                "user": {
                    "type": "person"
                },
                "field_name": {
                    "type": "string"
                },
                "field_value": {
                    "type": "string"
                }
            },
            "required": ["field_name", "field_value"]
        }
    }
}
```

I like this option better as we're more flexible for user needs. We might put ourselves into a corner if we're using JSON schema for validation but we're mapping specific sections as with the current approach.

Thoughts?

Reply:
> The dependencies are scaring me a little bit.
> 
> Also, I wonder how the JSON schema types are consistent with field types that are usable in document. If not the case, it could be a little bit weird for users.

After discussing with @afoucret on this, he mentioned to talk to core-infra team about the new dependency. Started [this slack thread](https://elastic.slack.com/archives/C8UUBNASY/p1681380940103139) with the conversation.